// @nativescript/google-maps/angular/fesm2022/nativescript-google-maps-angular.mjs
import * as i0 from '@angular/core';
import { Directive, NgModule } from '@angular/core';
import { registerElement } from '@nativescript/angular';
import { MapView } from '@nativescript/google-maps';

class MapViewDirective {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.4", ngImport: i0, type: MapViewDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.4", type: MapViewDirective, selector: "MapView", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.4", ngImport: i0, type: MapViewDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'MapView',
                }]
        }] });

class GoogleMapsModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.4", ngImport: i0, type: GoogleMapsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.1.4", ngImport: i0, type: GoogleMapsModule, declarations: [MapViewDirective], exports: [MapViewDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.1.4", ngImport: i0, type: GoogleMapsModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.4", ngImport: i0, type: GoogleMapsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MapViewDirective],
                    exports: [MapViewDirective],
                }]
        }] });
registerElement('MapView', () => MapView);

/**
 * Generated bundle index. Do not edit.
 */

export { GoogleMapsModule, MapViewDirective };
